File: benchmarks/cross-language/erlang/ring.erl
-module(ring).
-export([start/0, ring_actor/2]).

-define(RING_SIZE, 100).
-define(ROUNDS, 100000).
-define(MESSAGES, ?RING_SIZE * ?ROUNDS).

ring_actor(Next, Main) ->
    receive
        {token, 0} ->
            Main ! done;
        {token, Count} ->
            Next ! {token, Count - 1},
            ring_actor(Next, Main)
    end.

create_ring(0, First, Main) ->
    First ! {next, self(), Main},
    receive
        {next, Next, Main} -> Next
    end;
create_ring(N, First, Main) ->
    Next = create_ring(N - 1, First, Main),
    spawn(?MODULE, ring_actor, [Next, Main]).

start() ->
    io:format("=== Erlang Ring Benchmark ===~n"),
    io:format("Ring size: ~p~n", [?RING_SIZE]),
    io:format("Rounds: ~p~n", [?ROUNDS]),
    io:format("Total messages: ~p~n~n", [?MESSAGES]),

    Main = self(),

    % Create ring
    First = spawn(fun() ->
        receive
            {next, Next, M} ->
                ring_actor(Next, M)
        end
    end),

    _Last = create_ring(?RING_SIZE - 1, First, Main),

    % Start timer
    Start = erlang:system_time(nanosecond),

    % Send token
    First ! {token, ?MESSAGES},

    % Wait for completion
    receive
        done -> ok
    after 60000 ->
        io:format("Timeout!~n"),
        halt(1)
    end,

    % Calculate metrics
    End = erlang:system_time(nanosecond),
    ElapsedNs = End - Start,
    ElapsedSec = ElapsedNs / 1000000000.0,

    % Estimate cycles (assuming 3GHz)
    TotalCycles = ElapsedNs * 3.0,
    CyclesPerMsg = TotalCycles / ?MESSAGES,
    MsgPerSec = ?MESSAGES / ElapsedSec,

    io:format("Cycles/msg:     ~.2f~n", [CyclesPerMsg]),
    io:format("Throughput:     ~.0f M msg/sec~n", [MsgPerSec / 1000000]),

    halt(0).

File: benchmarks/cross-language/rust/src/ring.rs
use tokio::sync::mpsc;
use tokio::task;
use std::time::Instant;

const RING_SIZE: usize = 100;
const ROUNDS: i64 = 100_000;
const TOTAL_MESSAGES: i64 = RING_SIZE as i64 * ROUNDS;

async fn ring_actor(mut rx: mpsc::Receiver<i64>, tx: mpsc::Sender<i64>) {
    while let Some(token) = rx.recv().await {
        if token > 0 {
            let _ = tx.send(token - 1).await;
        } else {
            break;
        }
    }
}

#[tokio::main]
async fn main() {
    println!("=== Rust Ring Benchmark ===");
    println!("Ring size: {} actors", RING_SIZE);
    println!("Rounds: {}", ROUNDS);
    println!("Total messages: {}\n", TOTAL_MESSAGES);

    // Create ring of channels
    let (first_tx, mut first_rx) = mpsc::channel::<i64>(1000);
    let mut prev_tx = first_tx.clone();

    // Spawn ring actors
    for _ in 1..RING_SIZE {
        let (tx, rx) = mpsc::channel::<i64>(1000);
        let next_tx = tx.clone();
        task::spawn(ring_actor(rx, prev_tx));
        prev_tx = next_tx;
    }

    let start = Instant::now();

    // Send initial token
    let _ = prev_tx.send(TOTAL_MESSAGES).await;

    // Close loop: forward first_rx to prev_tx
    task::spawn(async move {
        while let Some(token) = first_rx.recv().await {
            if token > 0 {
                let _ = prev_tx.send(token - 1).await;
            }
        }
    });

    // Wait for completion
    tokio::time::sleep(tokio::time::Duration::from_secs(30)).await;

    let elapsed = start.elapsed();

    // Calculate metrics
    let cycles = (elapsed.as_secs_f64() * 3e9) as u64;
    let cycles_per_msg = cycles as f64 / TOTAL_MESSAGES as f64;
    let msg_per_sec = TOTAL_MESSAGES as f64 / elapsed.as_secs_f64();

    println!("Cycles/msg:     {:.2}", cycles_per_msg);
    println!("Throughput:     {:.0} M msg/sec", msg_per_sec / 1_000_000.0);
}
